### DISCUSSION TITLE
WatsonWebserver.Core.dll - System.Text.Json loading issue.

### DISCUSSION THREAD

**Discussion by GitHubProUser67** [dotnet/WatsonWebserver#151] [Q&A]
Hello, with the last version of the nuget package, an issue started to happen, which is related to assembly loading.

![image](https://github.com/dotnet/WatsonWebserver/assets/127040195/e28a62fb-63c4-4d4e-9afc-bf6ed38f7ad5)

This states that the assembly System.Text.Json is unable to be loaded.

I tried to copy it to the output directory, didn't help.

For some reasons this happens when loading it as a library alone, loading it within a server using the package makes this issue disapear.

So to sum it up:

HTTPServer -> Plugin -> WatsonWebserver.Core.dll -> Fails.

WatsonWebserver.Core.dll -> Costura.Fody -> HTTPSecureServer -> Plugin -> Success.

**Comment by jchristn** [comment]
Strange, this isn't happening for me using [WatsonWebserverTest](https://github.com/jchristn/watsonwebservertest).  Any additional steps to help reproduce?

**Comment by GitHubProUser67** [comment]
Yep, strange issue, the steps to reproduce are this:

Go to this project https://github.com/GitHubProUser67/MultiServer3

Dowload it via the download zip function, then press build.bat

After this go to ~BuildOutput/Multiserver/Debug|Release and start the HTTPServer.exe

**Comment by jchristn** [comment]
Just built it and didn't receive an error.

![image](https://github.com/dotnet/WatsonWebserver/assets/1421081/d081188c-c859-4f10-9c1b-26124b9dc0c0)

**Comment by GitHubProUser67** [comment]
Yep, that's because there is no plugins.

Ideally compile a plugin such as HomeWebTools and start the server.

**Comment by jchristn** [comment]
Sorry, I built from VS instead of running build.bat.  Reporting back in a moment.

**Comment by jchristn** [comment]
Yep, I see the exception.

It looks like you're running into an issue caused by dependency loading from using things like `Assembly.LoadFromFile` and other similar methods.

I've had to deal with this in other projects, and to get around it, I've had to ensure that subordinate projects correctly packed dependencies in their output.  Specifically, this in the .csproj file.  e.g. if project A is dynamically loading project B, project B must have this in the .csproj file, specifically within the top-level property group, to ensure the right DLLs are packed:

```
<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
<PublishSingleFile>true</PublishSingleFile>
<SelfContained>true</SelfContained>
```

I'm happy to chat with you to go through a very, very similar example to yours that caused me to go through this exact exercise and deal with this same problem.  Please feel free to email me at joel dot christner at gmail dot com.  

However, since this isn't an issue with the library, I'll close this and move it to discussions.  

Cheers, Joel

**Reply by GitHubProUser67** [reply]
Hello Joel, I want to thank you a lot for taking in charge my issue even if it is outside of the scope of the project.

So, I applied your recommandations for the backend section in it's entierty, and it worked!

![image](https://github.com/dotnet/WatsonWebserver/assets/127040195/e440ab70-dabe-4c83-9a53-e5dfda9d9677)

But the issue was actually more funny.

My server suit is designed to have a deps hierarchy that is custom. A plugin per say, can access the server code in a singular pipeline like this:

Plugin -> Server IPlugin -> Backend service.

In order when seeking a dependency, it will load each DLLs in the plugin section first, then, when loaded, those will "Seemlessly" (That's the problem) merge into the server code.

Problem is! Watson depends on the System.Text.Json library, but when looking for it, it will look in SERVER space.

So simply including the System.Text.Json on the HTTPServer itself and the csproj changes, fixed the problem.

Quite funny that a plugin loading optimization ended up doing this :).

**Comment by GitHubProUser67** [comment]
Closed in https://github.com/GitHubProUser67/MultiServer3/commit/3493666da237461e15eb8d0b165ebd353abbcbfb 

Many thanks to you Joel!

**Reply by jchristn** [reply]
Glad you got it resolved @GitHubProUser67 - if you ever want to chat about what I recently had to go through with a similar plug-in model, feel free to email me!  It sounds like a similar architecture.  Cheers, Joel

**Reply by GitHubProUser67** [reply]
I am interested in what you did with such a plugin system, I will email you when my schedule allows it.

Btw I did the "Real" fix with now allows to pack-in extra dependencies outside of server space, fixing properly the issue.

https://github.com/GitHubProUser67/MultiServer3/commit/fd2cc65baffd4154322f90992c30363b49554b31

I had to resort on Costura.Fody for this, as .NET single mode doesn't support my assembly loading system.
